File: src/main/scala/nl/rug/jbi/search/ssap/model/Instance.scala
package nl.rug.jbi.search.ssap.model

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.xml.Node

class Instance(val roles: mutable.Buffer[ScalaRole], val id:Long, val versions: mutable.Buffer[ScalaVersion]) {

  def toXml() = {
    <instance>
      {versions.map(_.toXml())}
      {roles.map(_.toXml())}
    </instance>
  }

  override def equals(o: Any) = o match {
    case that: Instance => that.roles.forall(r => roles.contains(r)) && that.roles.size == roles.size
    case that: Node =>
        val roleSeq = that \\ "role"
        roleSeq.forall(r => roles.contains(ScalaRole.fromXml(r))) && roleSeq.size == roles.size
    case _ => false
  }
  override def hashCode = roles.hashCode()
}

object Instance {
  private var currId:Long = -1

  def resetId() = currId = -1

  def fromXml(node: scala.xml.Node):Instance = {
    currId+=1
    val i:Instance = new Instance(ArrayBuffer[ScalaRole](), currId, ArrayBuffer[ScalaVersion]())
    (node \\ "role").map(r => i.roles+=ScalaRole.fromXml(r))
    (node \\ "version").map(v => i.versions+=ScalaVersion.fromXml(v))
    i
  }
}

File: ssap/src/main/java/nl/rug/jbi/search/ssap/model/Instance.java
package nl.rug.jbi.search.ssap.model;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamImplicit;
import com.thoughtworks.xstream.annotations.XStreamOmitField;

import java.util.HashSet;
import java.util.List;

@XStreamAlias("instance")
public class Instance {

    @XStreamOmitField
    private long id;

    @XStreamOmitField
    private static long currId = -1;

    @XStreamImplicit
    public List<Role> roleList;
    @XStreamImplicit
    private List<Version> versionList;

    private Instance (List<Role> roleList, List<Version> versionList) {
        this.roleList = roleList;
        currId++;
        this.id = currId;
        this.versionList = versionList;
    }

    public static void resetId () {
        currId = -1;
    }

    @Override
    public boolean equals (Object obj) {
        if (obj.getClass() == Instance.class) {
            return roleList.size() == ((Instance) obj).roleList.size() &&
                    new HashSet<>(roleList).containsAll(((Instance) obj).roleList);
        }
        return false;
    }

    @Override
    public int hashCode () {
        return roleList.hashCode();
    }
}
